Mathlib/LinearAlgebra/Span.lean

```diff
@@ -0,0 +1,1068 @@
+
+theorem sup_toAddSubgroup {R M : Type _} [Ring R] [AddCommGroup M] [Module R M]
+    (p p' : Submodule R M) : (p ⊔ p').toAddSubgroup = p.toAddSubgroup ⊔ p'.toAddSubgroup := by
+  ext x
+  rw [mem_toAddSubgroup, mem_sup, AddSubgroup.mem_sup]
+  rfl
+#align submodule.sup_to_add_subgroup Submodule.sup_toAddSubgroup
+
+end
+
+theorem mem_span_singleton_self (x : M) : x ∈ R ∙ x :=
+  subset_span rfl
+#align submodule.mem_span_singleton_self Submodule.mem_span_singleton_self
+
+theorem nontrivial_span_singleton {x : M} (h : x ≠ 0) : Nontrivial (R ∙ x) :=
+  ⟨by
+    use 0, ⟨x, Submodule.mem_span_singleton_self x⟩
+    intro H
+    rw [eq_comm, Submodule.mk_eq_zero] at H
+    exact h H⟩
+#align submodule.nontrivial_span_singleton Submodule.nontrivial_span_singleton
+
+theorem mem_span_singleton {y : M} : (x ∈ R ∙ y) ↔ ∃ a : R, a • y = x :=
+  ⟨fun h => by
+    refine span_induction h ?_ ?_ ?_ ?_
+    · rintro y (rfl | ⟨⟨_⟩⟩)
+      exact ⟨1, by simp⟩
+    · exact ⟨0, by simp⟩
+    · rintro _ _ ⟨a, rfl⟩ ⟨b, rfl⟩
+      exact ⟨a + b, by simp [add_smul]⟩
+    · rintro a _ ⟨b, rfl⟩
+      exact ⟨a * b, by simp [smul_smul]⟩, by
+    rintro ⟨a, y, rfl⟩; exact smul_mem _ _ (subset_span <| by simp)⟩
+#align submodule.mem_span_singleton Submodule.mem_span_singleton
+
+theorem le_span_singleton_iff {s : Submodule R M} {v₀ : M} :
+    (s ≤ R ∙ v₀) ↔ ∀ v ∈ s, ∃ r : R, r • v₀ = v := by simp_rw [SetLike.le_def, mem_span_singleton]
+#align submodule.le_span_singleton_iff Submodule.le_span_singleton_iff
+
+variable (R)
+
+theorem span_singleton_eq_top_iff (x : M) : (R ∙ x) = ⊤ ↔ ∀ v, ∃ r : R, r • x = v := by
+  rw [eq_top_iff, le_span_singleton_iff]
+  tauto
+#align submodule.span_singleton_eq_top_iff Submodule.span_singleton_eq_top_iff
+
+@[simp]
+theorem span_zero_singleton : (R ∙ (0 : M)) = ⊥ := by
+  ext
+  simp [mem_span_singleton, eq_comm]
+#align submodule.span_zero_singleton Submodule.span_zero_singleton
+
+theorem span_singleton_eq_range (y : M) : ↑(R ∙ y) = range ((· • y) : R → M) :=
+  Set.ext fun _ => mem_span_singleton
+#align submodule.span_singleton_eq_range Submodule.span_singleton_eq_range
+
+theorem span_singleton_smul_le {S} [Monoid S] [SMul S R] [MulAction S M] [IsScalarTower S R M]
+    (r : S) (x : M) : (R ∙ r • x) ≤ R ∙ x := by
+  rw [span_le, Set.singleton_subset_iff, SetLike.mem_coe]
+  exact smul_of_tower_mem _ _ (mem_span_singleton_self _)
+#align submodule.span_singleton_smul_le Submodule.span_singleton_smul_le
+
+theorem span_singleton_group_smul_eq {G} [Group G] [SMul G R] [MulAction G M] [IsScalarTower G R M]
+    (g : G) (x : M) : (R ∙ g • x) = R ∙ x := by
+  refine' le_antisymm (span_singleton_smul_le R g x) _
+  convert span_singleton_smul_le R g⁻¹ (g • x)
+  exact (inv_smul_smul g x).symm
+#align submodule.span_singleton_group_smul_eq Submodule.span_singleton_group_smul_eq
+
+variable {R}
+
+theorem span_singleton_smul_eq {r : R} (hr : IsUnit r) (x : M) : (R ∙ r • x) = R ∙ x := by
+  lift r to Rˣ using hr
```

Comment:
This complains with
```
./././Mathlib/LinearAlgebra/Span.lean:455:2: error: failed to synthesize
  CanLift R Rˣ ?m.157637 ?m.157638
```
because

https://github.com/leanprover-community/mathlib4/blob/1b617fcaa9246e0c51f0f6b9af7550c2af160f50/Mathlib/Algebra/Group/Units.lean#L585

I guess we should fix that porting note and add it back.